File: backend-dotnet/FinanceTracker.Api/Services/CategoryService.cs
using FinanceTracker.Api.Contracts;
using FinanceTracker.Api.Data;
using FinanceTracker.Api.Entities;
using FinanceTracker.Api.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace FinanceTracker.Api.Services;

public sealed class CategoryService(
    FinanceTrackerDbContext dbContext,
    CurrentUserService currentUserService,
    AccessControlLayer accessControlLayer,
    ILogger<CategoryService> logger)
{
    public async Task<IReadOnlyList<CategoryResponse>> GetAllAsync(CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var ownerIds = await accessControlLayer.GetAccessibleOwnerIdsAsync(userId, cancellationToken);

        var categories = await dbContext.Categories
            .AsNoTracking()
            .Where(category => ownerIds.Contains(category.UserId) && !category.Archived)
            .OrderBy(category => category.Name)
            .ToListAsync(cancellationToken);

        return categories.Select(ToResponse).ToList();
    }

    public async Task<CategoryResponse> CreateAsync(CategoryRequest request, CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var name = NormalizeName(request.name);

        await EnsureUniqueNameAsync(userId, request.type!.Value, name, null, cancellationToken);

        var category = new Category
        {
            UserId = userId,
            Name = name,
            Type = request.type!.Value,
            Color = request.color!,
            Icon = request.icon!,
            Archived = false
        };

        dbContext.Categories.Add(category);
        await dbContext.SaveChangesAsync(cancellationToken);
        return ToResponse(category);
    }

    public async Task<CategoryResponse> UpdateAsync(Guid id, CategoryRequest request, CancellationToken cancellationToken)
    {
        var userId = currentUserService.GetCurrentUserId();
        var category = await GetCategoryAsync(id, userId, cancellationToken);
        var name = NormalizeName(request.name);

        await EnsureUniqueNameAsync(userId, request.type!.Value, name, id, cancellationToken);

        category.Name = name;
        category.Type = request.type!.Value;
        category.Color = request.color!;
        category.Icon = request.icon!;

        await dbContext.SaveChangesAsync(cancellationToken);
        return ToResponse(category);
    }

    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken)
    {
        var category = await GetCategoryAsync(id, currentUserService.GetCurrentUserId(), cancellationToken);
        category.Archived = true;
        await dbContext.SaveChangesAsync(cancellationToken);
    }

    public async Task CreateDefaultCategoriesAsync(Guid userId, CancellationToken cancellationToken)
    {
        var trackedCategoryIds = dbContext.ChangeTracker.Entries<Category>()
            .Where(entry => entry.Entity.Id != Guid.Empty)
            .Select(entry => entry.Entity.Id)
            .Distinct()
            .ToList();

        if (trackedCategoryIds.Count > 0)
        {
            logger.LogInformation(
                "Default category creation starting with {TrackedCount} tracked categories for user {UserId}: {CategoryIds}",
                trackedCategoryIds.Count,
                userId,
                string.Join(", ", trackedCategoryIds));
        }

        var hasCategories = await dbContext.Categories
            .AsNoTracking()
            .AnyAsync(
            category => category.UserId == userId && !category.Archived,
            cancellationToken);

        if (hasCategories)
        {
            logger.LogInformation("Skipping default category creation because categories already exist for user {UserId}", userId);
            return;
        }

        var categories = DefaultExpenseCategories()
            .Select(item => BuildDefaultCategory(userId, item.name, CategoryType.EXPENSE, item.color, item.icon))
            .Concat(DefaultIncomeCategories()
                .Select(item => BuildDefaultCategory(userId, item.name, CategoryType.INCOME, item.color, item.icon)))
            .ToList();

        logger.LogInformation(
            "Creating {CategoryCount} default categories for user {UserId}: {Categories}",
            categories.Count,
            userId,
            string.Join(", ", categories.Select(category => $"{category.Id}:{category.Name}:{category.Type}")));

        dbContext.Categories.AddRange(categories);
        await dbContext.SaveChangesAsync(cancellationToken);
    }

    public async Task<Category> GetCategoryAsync(Guid id, Guid userId, CancellationToken cancellationToken)
    {
        return await dbContext.Categories
            .FirstOrDefaultAsync(category => category.Id == id && category.UserId == userId && !category.Archived, cancellationToken)
               ?? throw new NotFoundException("Category not found");
    }

    public async Task<Category> GetCategoryForAccountAsync(Guid id, Guid accountOwnerUserId, CancellationToken cancellationToken)
    {
        return await dbContext.Categories
            .FirstOrDefaultAsync(
                category => category.Id == id
                            && category.UserId == accountOwnerUserId
                            && !category.Archived,
                cancellationToken)
               ?? throw new NotFoundException("Category not found");
    }

    private static Category BuildDefaultCategory(Guid userId, string name, CategoryType type, string color, string icon)
    {
        return new Category
        {
            UserId = userId,
            Name = name,
            Type = type,
            Color = color,
            Icon = icon,
            Archived = false
        };
    }

    private async Task EnsureUniqueNameAsync(
        Guid userId,
        CategoryType type,
        string normalizedName,
        Guid? existingCategoryId,
        CancellationToken cancellationToken)
    {
        var exists = await dbContext.Categories
            .AsNoTracking()
            .AnyAsync(
                category => category.UserId == userId
                            && category.Type == type
                            && category.Name.ToLower() == normalizedName.ToLower()
                            && (existingCategoryId == null || category.Id != existingCategoryId.Value),
                cancellationToken);

        if (exists)
        {
            throw new BadRequestException("Category already exists");
        }
    }

    private static string NormalizeName(string? name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            throw new BadRequestException("Category name must not be blank");
        }

        return name.Trim();
    }

    private static IReadOnlyList<(string name, string color, string icon)> DefaultExpenseCategories()
    {
        return
        [
            ("Food", "#EF4444", "utensils"),
            ("Rent", "#F97316", "house"),
            ("Utilities", "#EAB308", "bolt"),
            ("Transport", "#14B8A6", "car"),
            ("Entertainment", "#8B5CF6", "film"),
            ("Shopping", "#EC4899", "shopping-bag"),
            ("Health", "#10B981", "heart-pulse"),
            ("Education", "#3B82F6", "graduation-cap"),
            ("Travel", "#06B6D4", "plane"),
            ("Subscriptions", "#6366F1", "repeat"),
            ("Misc", "#6B7280", "circle")
        ];
    }

    private static IReadOnlyList<(string name, string color, string icon)> DefaultIncomeCategories()
    {
        return
        [
            ("Salary", "#16A34A", "briefcase"),
            ("Freelance", "#0EA5E9", "laptop"),
            ("Bonus", "#84CC16", "sparkles"),
            ("Investment", "#22C55E", "chart-line"),
            ("Gift", "#F59E0B", "gift"),
            ("Refund", "#06B6D4", "rotate-ccw"),
            ("Other", "#64748B", "plus-circle")
        ];
    }

    private static CategoryResponse ToResponse(Category category)
    {
        return new CategoryResponse(
            category.Id,
            category.Name,
            category.Type,
            category.Color,
            category.Icon,
            category.Archived
        );
    }
}

File: backend/src/main/java/com/finance/services/CategoryService.java
package com.finance.services;

import com.finance.dto.CategoryDtos.CategoryRequest;
import com.finance.dto.CategoryDtos.CategoryResponse;
import com.finance.entities.Category;
import com.finance.entities.CategoryType;
import com.finance.entities.User;
import com.finance.exception.BadRequestException;
import com.finance.exception.NotFoundException;
import com.finance.repositories.CategoryRepository;
import com.finance.repositories.UserRepository;
import com.finance.security.CurrentUserService;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;
    private final UserRepository userRepository;
    private final CurrentUserService currentUserService;

    @Transactional(readOnly = true)
    public List<CategoryResponse> getAll() {
        return categoryRepository.findByUserIdAndArchivedFalseOrderByNameAsc(currentUserService.getCurrentUserId())
                .stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional
    public CategoryResponse create(CategoryRequest request) {
        UUID userId = currentUserService.getCurrentUserId();
        if (categoryRepository.existsByUserIdAndNameIgnoreCaseAndType(userId, request.name(), request.type())) {
            throw new BadRequestException("Category already exists");
        }

        Category category = new Category();
        category.setUser(getUser(userId));
        map(category, request);
        category.setArchived(false);
        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public CategoryResponse update(UUID id, CategoryRequest request) {
        Category category = getCategory(id, currentUserService.getCurrentUserId());
        map(category, request);
        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public void delete(UUID id) {
        Category category = getCategory(id, currentUserService.getCurrentUserId());
        category.setArchived(true);
        categoryRepository.save(category);
    }

    @Transactional
    public void createDefaultCategories(User user) {
        if (!categoryRepository.findByUserIdAndArchivedFalseOrderByNameAsc(user.getId()).isEmpty()) {
            return;
        }

        List<Category> categories = new ArrayList<>();
        defaultExpenseCategories().forEach((name, meta) -> categories.add(buildDefaultCategory(user, name, CategoryType.EXPENSE, meta[0], meta[1])));
        defaultIncomeCategories().forEach((name, meta) -> categories.add(buildDefaultCategory(user, name, CategoryType.INCOME, meta[0], meta[1])));
        categoryRepository.saveAll(categories);
    }

    @Transactional(readOnly = true)
    public Category getCategory(UUID id, UUID userId) {
        return categoryRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new NotFoundException("Category not found"));
    }

    private User getUser(UUID userId) {
        return userRepository.findById(userId).orElseThrow(() -> new NotFoundException("User not found"));
    }

    private void map(Category category, CategoryRequest request) {
        category.setName(request.name());
        category.setType(request.type());
        category.setColor(request.color());
        category.setIcon(request.icon());
    }

    private Category buildDefaultCategory(User user, String name, CategoryType type, String color, String icon) {
        Category category = new Category();
        category.setUser(user);
        category.setName(name);
        category.setType(type);
        category.setColor(color);
        category.setIcon(icon);
        category.setArchived(false);
        return category;
    }

    private Map<String, String[]> defaultExpenseCategories() {
        return Map.ofEntries(
                Map.entry("Food", new String[]{"#EF4444", "utensils"}),
                Map.entry("Rent", new String[]{"#F97316", "house"}),
                Map.entry("Utilities", new String[]{"#EAB308", "bolt"}),
                Map.entry("Transport", new String[]{"#14B8A6", "car"}),
                Map.entry("Entertainment", new String[]{"#8B5CF6", "film"}),
                Map.entry("Shopping", new String[]{"#EC4899", "shopping-bag"}),
                Map.entry("Health", new String[]{"#10B981", "heart-pulse"}),
                Map.entry("Education", new String[]{"#3B82F6", "graduation-cap"}),
                Map.entry("Travel", new String[]{"#06B6D4", "plane"}),
                Map.entry("Subscriptions", new String[]{"#6366F1", "repeat"}),
                Map.entry("Misc", new String[]{"#6B7280", "circle"})
        );
    }

    private Map<String, String[]> defaultIncomeCategories() {
        return Map.ofEntries(
                Map.entry("Salary", new String[]{"#16A34A", "briefcase"}),
                Map.entry("Freelance", new String[]{"#0EA5E9", "laptop"}),
                Map.entry("Bonus", new String[]{"#84CC16", "sparkles"}),
                Map.entry("Investment", new String[]{"#22C55E", "chart-line"}),
                Map.entry("Gift", new String[]{"#F59E0B", "gift"}),
                Map.entry("Refund", new String[]{"#06B6D4", "rotate-ccw"}),
                Map.entry("Other", new String[]{"#64748B", "plus-circle"})
        );
    }

    private CategoryResponse toResponse(Category category) {
        return new CategoryResponse(
                category.getId(),
                category.getName(),
                category.getType(),
                category.getColor(),
                category.getIcon(),
                category.isArchived()
        );
    }
}
